lua/diffview/vcs/adapters/hg/init.lua

```diff
@@ -1,18 +1,1141 @@
+  ---@type LogEntry[]
+  local entries = {}
+  local data = {}
+  local data_idx = 1
+  local last_status
+  local err_msg
+
+  local single_file = self:is_single_file(log_opt.single_file.path_args, {})
+
+  ---@type LogOptions
+  local log_options = config.get_log_options(
+    single_file,
+    single_file and log_opt.single_file or log_opt.multi_file,
+    "hg"
+  )
+
+  local state = {
+    thread = thread,
+    adapter = self,
+    path_args = log_opt.single_file.path_args,
+    log_options = log_options,
+    prepared_log_opts = prepare_fh_options(self, log_options, single_file),
+    opt = opt,
+    callback = callback,
+    entries = entries,
+    single_file = single_file,
+    resume_lock = false,
+  }
+
+  local function data_callback(status, d, msg)
+    if status == JobStatus.PROGRESS then
+      data[#data+1] = d
+    end
+
+    last_status = status
+    if msg then
+      err_msg = msg
+    end
+    if not state.resume_lock and coroutine.status(thread) == "suspended" then
+      self:handle_co(thread, coroutine.resume(thread))
+    end
+
+    if co_state.shutdown then
+      return true
+    end
+  end
+
+  incremental_fh_data(state, data_callback)
+
+  while true do
+    if not vim.tbl_contains({ JobStatus.SUCCESS, JobStatus.ERROR, JobStatus.KILLED }, last_status)
+        and not data[data_idx] then
+      coroutine.yield()
+    end
+
+    if last_status == JobStatus.KILLED then
+      logger.warn("File history processing was killed.")
+      return
+    elseif last_status == JobStatus.ERROR then
+      callback(entries, JobStatus.ERROR, err_msg)
+      return
+    elseif last_status == JobStatus.SUCCESS and data_idx > #data then
+      break
+    end
+
+    state.cur = data[data_idx]
+
+    state.commit = Commit({
+      hash = state.cur.right_hash,
+      author = state.cur.author,
+      time = tonumber(state.cur.time),
+      time_offset = state.cur.time_offset,
+      rel_date = state.cur.rel_date,
+      ref_names = state.cur.ref_names,
+      subject = state.cur.subject,
+    })
+
+    local ok, status = parse_fh_data(state)
+
+    if not ok then
+      if status == JobStatus.FATAL then
+        return
+      end
+      break
+    end
+
+    data_idx = data_idx + 1
+  end
+
+  callback(entries, JobStatus.SUCCESS)
+end
+
+function HgAdapter:diffview_options(args)
+  local default_args = config.get_config().default_args.DiffviewOpen
+  local argo = arg_parser.parse(vim.tbl_flatten({ default_args, args }))
+  local rev_args = argo:get_flag({'rev'})
+
+  local head = self:head_rev()
+  local left = head or HgRev.new_null_tree()
+  local right = HgRev(RevType.LOCAL)
+
+  local options = {
+    show_untracked = true, -- TODO: extract from hg config
+    selected_file = argo:get_flag("selected-file", { no_empty = true, expand = true })
+      or (vim.bo.buftype == "" and pl:vim_expand("%:p"))
+      or nil,
+  }
+
+  return {left = left, right = right, options = options}
+end
```

Comment:
The revs here are hardcoded and do not respect the rev argument passed by the user. The rev arg should be the first non-option argument, which can be accessed like this: `argo.args[1]`. Should probably implement something like the `GitAdapter:parse_revs()` method.

Reply:
In hg the rev is always passed via `-r <revset>` (or `--rev=<revset>`), so not as the unnamed argument like it is in git.

In all other cases we use the same command options and flags, so I thought it makes sense to do that here too?

Reply:
It's not as clear cut as that. `:DiffviewOpen` isn't directly synonymous with `git diff`. It was designed to be similar to the git command of course, but it has its own command structure. For instance the git command allows you to specify the range as two separate rev args, our command does not. The git command generally allows you to specify paths without the end-of-options token (`--`) as long as the args are not ambiguous. Our command does not (to avoid dealing with arg ambiguity). And finally, our command has some custom options that are not from git. I kind of regret not naming these options differently such that it would be clear that they are not git options, but oh well, that's off-topic.

Point is, `:DiffviewOpen` is it's own thing, and it has its own command structure that I would like to stick to:

```
:DiffviewOpen [rev] [options] [-- [paths...]]
```

Reply:
Alright, I'll fix it